zkevm-circuits/src/evm_circuit/op_execution.rs

```diff
@@ -0,0 +1,532 @@
+
+        let mut begin_idx = num_case;
+        let mut rotation_range = Range::default();
+        for (idx, cell) in free_cells.iter().enumerate() {
+            // skip case selector
+            if idx < num_case {
+                continue;
+            }
+
+            let next_idx = idx + 1;
+            if next_idx == num_free_cell || word_ranges.len() == self.num_word {
+                cell_idxs.append(&mut (idx..num_free_cell).collect());
+                break;
+            }
+
+            // find cells in rectangle region for words
+            if rotation_range.contains(&cell.rotation) {
+                if next_idx - begin_idx == 32 {
+                    let next_cell = &free_cells[next_idx];
+                    word_ranges.push(begin_idx..next_idx);
+                    begin_idx = next_idx;
+                    rotation_range =
+                        next_cell.rotation..next_cell.rotation + word_height;
+                }
+            } else {
+                cell_idxs.append(&mut (begin_idx..idx).collect());
+                begin_idx = idx;
+                rotation_range = cell.rotation..cell.rotation + word_height;
+            }
+        }
+
+        assert_eq!(
+            word_ranges.len(),
+            self.num_word,
+            "not enough rectangle regions for words"
+        );
+
+        let (cell_idxs, unused_idxs) = cell_idxs.split_at(self.num_cell);
+        (word_ranges, cell_idxs.to_vec(), unused_idxs.to_vec())
+    }
+}
+
+#[derive(Debug)]
+struct CaseAllocation<F> {
+    selector: Cell<F>,
+    words: Vec<Word<F>>,
+    cells: Vec<Cell<F>>,
+    resumption: Option<Resumption<F>>,
+}
+
+#[derive(Clone, Debug)]
+struct Resumption<F> {
+    caller_id: Cell<F>,
+    gas_available: Cell<F>,
+}
+
+impl<F: FieldExt> Resumption<F> {
+    fn new(cells: &[Cell<F>]) -> Self {
+        assert_eq!(cells.len(), NUM_CELL_RESUMPTION);
+
+        Self {
+            caller_id: cells[0].clone(),
+            gas_available: cells[1].clone(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct OpExecutionState<F> {
+    pub is_executing: Cell<F>,
+    pub global_counter: Cell<F>,
+    pub call_id: Cell<F>,
+    pub program_counter: Cell<F>,
+    pub stack_pointer: Cell<F>,
+    gas_counter: Cell<F>,
+    opcode: Cell<F>,
+}
+
+impl<F: FieldExt> OpExecutionState<F> {
+    pub(crate) fn new(cells: &[Cell<F>]) -> Self {
+        assert_eq!(cells.len(), NUM_CELL_OP_EXECTION_STATE);
+
+        Self {
+            is_executing: cells[0].clone(),
+            global_counter: cells[1].clone(),
+            call_id: cells[2].clone(),
+            program_counter: cells[3].clone(),
+            stack_pointer: cells[4].clone(),
+            gas_counter: cells[5].clone(),
+            opcode: cells[6].clone(),
+        }
+    }
+}
+
+trait OpGadget<F: FieldExt> {
+    const RESPONSIBLE_OPCODES: &'static [u8];
+
+    const CASE_CONFIGS: &'static [CaseConfig];
```

Comment:
Should this be provided by an `ExecutionStep` or similar? Also, I don't really see how is possible to make this a const if it changes for every opcode. And between opcode instances.

Reply:
But this is only for circuit constraint configuration usage, it tells how many cases this `OpGadget` could encounter, and `OpExecutionGadget` will allocate enough cells for it.

When prover is assigning witness, the `OpGadget` will assign correct values depends on the case provided by `ExecutionStep`.

Reply:
>But this is only for circuit constraint configuration usage, it tells how many cases this OpGadget could encounter, and 
> OpExecutionGadget will allocate enough cells for it.

This sounds  more to me like an enum than rather an slice. But if you think that is the most practical way, I'm fine for now! :)

Reply:
Ah, I think I got your point, you mean we should separate each case as a unique `OpGadget`?

Reply:
Exactly. But you'll know better what fits there.